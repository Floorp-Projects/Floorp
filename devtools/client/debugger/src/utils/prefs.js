/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

const { PrefsHelper } = require("devtools/client/shared/prefs");

import { isNode } from "./environment";
import Services from "devtools-services";

// Schema version to bump when the async store format has changed incompatibly
// and old stores should be cleared.
const prefsSchemaVersion = 11;
const { pref } = Services;

if (isNode()) {
  pref("devtools.browsertoolbox.fission", false);
  pref("devtools.debugger.logging", false);
  pref("devtools.debugger.alphabetize-outline", false);
  pref("devtools.debugger.auto-pretty-print", false);
  pref("devtools.source-map.client-service.enabled", true);
  pref("devtools.chrome.enabled", false);
  pref("devtools.debugger.pause-on-exceptions", false);
  pref("devtools.debugger.pause-on-caught-exceptions", false);
  pref("devtools.debugger.ignore-caught-exceptions", true);
  pref("devtools.debugger.call-stack-visible", true);
  pref("devtools.debugger.scopes-visible", true);
  pref("devtools.debugger.component-visible", false);
  pref("devtools.debugger.workers-visible", false);
  pref("devtools.debugger.expressions-visible", false);
  pref("devtools.debugger.xhr-breakpoints-visible", false);
  pref("devtools.debugger.breakpoints-visible", true);
  pref("devtools.debugger.event-listeners-visible", false);
  pref("devtools.debugger.dom-mutation-breakpoints-visible", false);
  pref("devtools.debugger.start-panel-collapsed", false);
  pref("devtools.debugger.end-panel-collapsed", false);
  pref("devtools.debugger.start-panel-size", 300);
  pref("devtools.debugger.end-panel-size", 300);
  pref("devtools.debugger.tabsBlackBoxed", "[]");
  pref("devtools.debugger.ui.editor-wrapping", false);
  pref("devtools.debugger.ui.framework-grouping-on", true);
  pref("devtools.debugger.pending-selected-location", "{}");
  pref("devtools.debugger.expressions", "[]");
  pref("devtools.debugger.file-search-case-sensitive", false);
  pref("devtools.debugger.file-search-whole-word", false);
  pref("devtools.debugger.file-search-regex-match", false);
  pref("devtools.debugger.project-directory-root", "");
  pref("devtools.debugger.map-scopes-enabled", false);
  pref("devtools.debugger.prefs-schema-version", prefsSchemaVersion);
  pref("devtools.debugger.skip-pausing", false);
  pref("devtools.debugger.log-actions", true);
  pref("devtools.debugger.log-event-breakpoints", false);
  pref("devtools.debugger.features.workers", true);
  pref("devtools.debugger.features.async-stepping", false);
  pref("devtools.debugger.features.wasm", true);
  pref("devtools.debugger.features.shortcuts", true);
  pref("devtools.debugger.features.root", true);
  pref("devtools.debugger.features.map-scopes", true);
  pref("devtools.debugger.features.remove-command-bar-options", true);
  pref("devtools.debugger.features.code-folding", false);
  pref("devtools.debugger.features.command-click", false);
  pref("devtools.debugger.features.outline", true);
  pref("devtools.debugger.features.column-breakpoints", true);
  pref("devtools.debugger.features.skip-pausing", true);
  pref("devtools.debugger.features.component-pane", false);
  pref("devtools.debugger.features.autocomplete-expressions", false);
  pref("devtools.debugger.features.map-expression-bindings", true);
  pref("devtools.debugger.features.map-await-expression", true);
  pref("devtools.debugger.features.xhr-breakpoints", true);
  pref("devtools.debugger.features.original-blackbox", true);
  pref("devtools.debugger.features.event-listeners-breakpoints", true);
  pref("devtools.debugger.features.dom-mutation-breakpoints", true);
  pref("devtools.debugger.features.log-points", true);
  pref("devtools.debugger.features.inline-preview", true);
  pref("devtools.debugger.features.overlay-step-buttons", true);
  pref("devtools.debugger.features.frame-step", true);
  pref("devtools.editor.tabsize", 2);
}

export const prefs = new PrefsHelper("devtools", {
  fission: ["Bool", "browsertoolbox.fission"],
  logging: ["Bool", "debugger.logging"],
  editorWrapping: ["Bool", "debugger.ui.editor-wrapping"],
  alphabetizeOutline: ["Bool", "debugger.alphabetize-outline"],
  autoPrettyPrint: ["Bool", "debugger.auto-pretty-print"],
  clientSourceMapsEnabled: ["Bool", "source-map.client-service.enabled"],
  chromeAndExtensionsEnabled: ["Bool", "chrome.enabled"],
  pauseOnExceptions: ["Bool", "debugger.pause-on-exceptions"],
  pauseOnCaughtExceptions: ["Bool", "debugger.pause-on-caught-exceptions"],
  ignoreCaughtExceptions: ["Bool", "debugger.ignore-caught-exceptions"],
  callStackVisible: ["Bool", "debugger.call-stack-visible"],
  scopesVisible: ["Bool", "debugger.scopes-visible"],
  componentVisible: ["Bool", "debugger.component-visible"],
  workersVisible: ["Bool", "debugger.workers-visible"],
  breakpointsVisible: ["Bool", "debugger.breakpoints-visible"],
  expressionsVisible: ["Bool", "debugger.expressions-visible"],
  xhrBreakpointsVisible: ["Bool", "debugger.xhr-breakpoints-visible"],
  eventListenersVisible: ["Bool", "debugger.event-listeners-visible"],
  domMutationBreakpointsVisible: [
    "Bool",
    "debugger.dom-mutation-breakpoints-visible",
  ],
  startPanelCollapsed: ["Bool", "debugger.start-panel-collapsed"],
  endPanelCollapsed: ["Bool", "debugger.end-panel-collapsed"],
  startPanelSize: ["Int", "debugger.start-panel-size"],
  endPanelSize: ["Int", "debugger.end-panel-size"],
  frameworkGroupingOn: ["Bool", "debugger.ui.framework-grouping-on"],
  pendingSelectedLocation: ["Json", "debugger.pending-selected-location", {}],
  expressions: ["Json", "debugger.expressions", []],
  fileSearchCaseSensitive: ["Bool", "debugger.file-search-case-sensitive"],
  fileSearchWholeWord: ["Bool", "debugger.file-search-whole-word"],
  fileSearchRegexMatch: ["Bool", "debugger.file-search-regex-match"],
  debuggerPrefsSchemaVersion: ["Int", "debugger.prefs-schema-version"],
  projectDirectoryRoot: ["Char", "debugger.project-directory-root", ""],
  projectDirectoryRootName: [
    "Char",
    "debugger.project-directory-root-name",
    "",
  ],
  skipPausing: ["Bool", "debugger.skip-pausing"],
  mapScopes: ["Bool", "debugger.map-scopes-enabled"],
  logActions: ["Bool", "debugger.log-actions"],
  logEventBreakpoints: ["Bool", "debugger.log-event-breakpoints"],
  indentSize: ["Int", "editor.tabsize"],
});

// The pref may not be defined. Defaulting to null isn't viable (cursor never blinks).
// Can't use CodeMirror.defaults here because it's loaded later.
// Hardcode the fallback value to that of CodeMirror.defaults.cursorBlinkRate.
prefs.cursorBlinkRate = Services.prefs.getIntPref("ui.caretBlinkTime", 530);

export const features = new PrefsHelper("devtools.debugger.features", {
  asyncStepping: ["Bool", "async-stepping"],
  wasm: ["Bool", "wasm"],
  shortcuts: ["Bool", "shortcuts"],
  root: ["Bool", "root"],
  columnBreakpoints: ["Bool", "column-breakpoints"],
  mapScopes: ["Bool", "map-scopes"],
  removeCommandBarOptions: ["Bool", "remove-command-bar-options"],
  workers: ["Bool", "workers"],
  windowlessWorkers: ["Bool", "windowless-workers"],
  outline: ["Bool", "outline"],
  codeFolding: ["Bool", "code-folding"],
  skipPausing: ["Bool", "skip-pausing"],
  autocompleteExpression: ["Bool", "autocomplete-expressions"],
  mapExpressionBindings: ["Bool", "map-expression-bindings"],
  mapAwaitExpression: ["Bool", "map-await-expression"],
  componentPane: ["Bool", "component-pane"],
  xhrBreakpoints: ["Bool", "xhr-breakpoints"],
  originalBlackbox: ["Bool", "original-blackbox"],
  eventListenersBreakpoints: ["Bool", "event-listeners-breakpoints"],
  domMutationBreakpoints: ["Bool", "dom-mutation-breakpoints"],
  logPoints: ["Bool", "log-points"],
  commandClick: ["Bool", "command-click"],
  showOverlay: ["Bool", "overlay"],
  inlinePreview: ["Bool", "inline-preview"],
  windowlessServiceWorkers: ["Bool", "windowless-service-workers"],
  frameStep: ["Bool", "frame-step"],
});

// Import the asyncStore already spawned by the TargetMixin class
const ThreadUtils = require("devtools/client/shared/thread-utils");
export const asyncStore = ThreadUtils.asyncStore;

export function resetSchemaVersion() {
  prefs.debuggerPrefsSchemaVersion = prefsSchemaVersion;
}

export function verifyPrefSchema() {
  if (prefs.debuggerPrefsSchemaVersion < prefsSchemaVersion) {
    asyncStore.pendingBreakpoints = {};
    asyncStore.tabs = [];
    asyncStore.xhrBreakpoints = [];
    asyncStore.eventListenerBreakpoints = undefined;
    prefs.debuggerPrefsSchemaVersion = prefsSchemaVersion;
  }
}
