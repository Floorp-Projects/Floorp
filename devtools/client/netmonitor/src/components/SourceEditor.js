/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { Component } = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  connect,
} = require("devtools/client/shared/redux/visibility-handler-connect");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const Editor = require("devtools/client/shared/sourceeditor/editor");
const {
  setTargetSearchResult,
} = require("devtools/client/netmonitor/src/actions/search");
const { div } = dom;

/**
 * CodeMirror editor as a React component
 */
class SourceEditor extends Component {
  static get propTypes() {
    return {
      // Source editor syntax highlight mode, which is a mime type defined in CodeMirror
      mode: PropTypes.string,
      // Source editor content
      text: PropTypes.string,
      // Auto scroll to specific line
      scrollToLine: PropTypes.number,
      // Reset target search result that has been used for navigation in this panel.
      // This is done to avoid second navigation the next time.
      resetTargetSearchResult: PropTypes.func,
    };
  }

  componentDidMount() {
    const { mode, text } = this.props;

    this.editor = new Editor({
      lineNumbers: true,
      lineWrapping: false,
      mode: null, // Disable auto syntax detection, but then we set mode asynchronously
      readOnly: true,
      theme: "mozilla",
      value: text,
    });

    // Delay to CodeMirror initialization content to prevent UI freezing
    this.editorTimeout = setTimeout(() => {
      this.editorTimeout = null;
      this.editor.appendToLocalElement(this.refs.editorElement);

      // CodeMirror's setMode() (syntax highlight) is the performance bottleneck when
      // processing large content, so we enable it asynchronously within the setTimeout
      // to avoid UI blocking. (rendering source code -> drawing syntax highlight)
      this.editorSetModeTimeout = setTimeout(() => {
        this.editorSetModeTimeout = null;
        this.editor.setMode(mode);
        this.scrollToLine();
      });
    });
  }

  shouldComponentUpdate(nextProps) {
    return (
      nextProps.mode !== this.props.mode ||
      nextProps.text !== this.props.text ||
      nextProps.scrollToLine !== this.props.scrollToLine
    );
  }

  componentDidUpdate(prevProps) {
    const { mode, scrollToLine, text } = this.props;

    // Bail out if the editor has been destroyed in the meantime.
    if (this.editor.isDestroyed()) {
      return;
    }

    if (prevProps.text !== text) {
      // Reset the existed 'mode' attribute in order to make setText() process faster
      // to prevent drawing unnecessary syntax highlight.
      this.editor.setMode(null);
      this.editor.setText(text);

      if (this.editorSetModeTimeout) {
        clearTimeout(this.editorSetModeTimeout);
      }

      // CodeMirror's setMode() (syntax highlight) is the performance bottleneck when
      // processing large content, so we enable it asynchronously within the setTimeout
      // to avoid UI blocking. (rendering source code -> drawing syntax highlight)
      this.editorSetModeTimeout = setTimeout(() => {
        this.editorSetModeTimeout = null;
        this.editor.setMode(mode);
        this.scrollToLine();
      });
    } else if (prevProps.scrollToLine !== scrollToLine) {
      this.scrollToLine();
    }
  }

  componentWillUnmount() {
    clearTimeout(this.editorTimeout);
    clearTimeout(this.editorSetModeTimeout);
    this.editor.destroy();
  }

  scrollToLine() {
    const { scrollToLine, resetTargetSearchResult } = this.props;

    if (scrollToLine) {
      this.editor.setCursor(
        {
          line: scrollToLine - 1,
        },
        "center"
      );
    }

    resetTargetSearchResult();
  }

  render() {
    return div({
      ref: "editorElement",
      className: "source-editor-mount devtools-monospace",
    });
  }
}

module.exports = connect(null, dispatch => ({
  resetTargetSearchResult: () => dispatch(setTargetSearchResult(null)),
}))(SourceEditor);
