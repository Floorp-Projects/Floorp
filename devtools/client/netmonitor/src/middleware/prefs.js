/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const Services = require("Services");
const {
  ENABLE_REQUEST_FILTER_TYPE_ONLY,
  RESET_COLUMNS,
  TOGGLE_COLUMN,
  TOGGLE_REQUEST_FILTER_TYPE,
  ENABLE_PERSISTENT_LOGS,
  DISABLE_BROWSER_CACHE,
  SET_COLUMNS_WIDTH,
  WS_TOGGLE_COLUMN,
  WS_RESET_COLUMNS,
} = require("devtools/client/netmonitor/src/constants");

/**
 * Update the relevant prefs when:
 *   - a column has been toggled
 *   - a filter type has been set
 */
function prefsMiddleware(store) {
  return next => action => {
    const res = next(action);
    switch (action.type) {
      case ENABLE_REQUEST_FILTER_TYPE_ONLY:
      case TOGGLE_REQUEST_FILTER_TYPE:
        const filters = Object.entries(
          store.getState().filters.requestFilterTypes
        )
          .filter(([type, check]) => check)
          .map(([type, check]) => type);
        Services.prefs.setCharPref(
          "devtools.netmonitor.filters",
          JSON.stringify(filters)
        );
        break;
      case ENABLE_PERSISTENT_LOGS:
        const enabled = store.getState().ui.persistentLogsEnabled;
        Services.prefs.setBoolPref("devtools.netmonitor.persistlog", enabled);
        break;
      case DISABLE_BROWSER_CACHE:
        Services.prefs.setBoolPref(
          "devtools.cache.disabled",
          store.getState().ui.browserCacheDisabled
        );
        break;
      case TOGGLE_COLUMN:
        persistVisibleColumns(store.getState());
        break;
      case RESET_COLUMNS:
        persistVisibleColumns(store.getState());
        persistColumnsData(store.getState());
        break;
      case SET_COLUMNS_WIDTH:
        persistColumnsData(store.getState());
        break;
      case WS_TOGGLE_COLUMN:
      case WS_RESET_COLUMNS:
        persistVisibleWebSocketsColumns(store.getState());
        break;
    }
    return res;
  };
}

/**
 * Store list of visible columns into preferences.
 */
function persistVisibleColumns(state) {
  const visibleColumns = [];
  const { columns } = state.ui;
  for (const column in columns) {
    if (columns[column]) {
      visibleColumns.push(column);
    }
  }

  Services.prefs.setCharPref(
    "devtools.netmonitor.visibleColumns",
    JSON.stringify(visibleColumns)
  );
}

/**
 * Store list of visible columns into preferences.
 */
function persistVisibleWebSocketsColumns(state) {
  const visibleColumns = [];
  const { columns } = state.messages;
  for (const column in columns) {
    if (columns[column]) {
      visibleColumns.push(column);
    }
  }

  Services.prefs.setCharPref(
    "devtools.netmonitor.msg.visibleColumns",
    JSON.stringify(visibleColumns)
  );
}

/**
 * Store columns data (width, min-width, etc.) into preferences.
 */
function persistColumnsData(state) {
  const columnsData = [...state.ui.columnsData.values()];
  Services.prefs.setCharPref(
    "devtools.netmonitor.columnsData",
    JSON.stringify(columnsData)
  );
}

module.exports = prefsMiddleware;
