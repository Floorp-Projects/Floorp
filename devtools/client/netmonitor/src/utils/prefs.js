/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { PrefsHelper } = require("devtools/client/shared/prefs");

/**
 * Shortcuts for accessing various network monitor preferences.
 */
exports.Prefs = new PrefsHelper("devtools.netmonitor", {
  networkDetailsWidth: ["Int", "panes-network-details-width"],
  networkDetailsHeight: ["Int", "panes-network-details-height"],
  visibleColumns: ["Json", "visibleColumns"],
  columnsData: ["Json", "columnsData"],
  filters: ["Json", "filters"],
});
