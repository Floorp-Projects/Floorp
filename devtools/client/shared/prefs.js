/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
"use strict";

const Services = require("Services");
const EventEmitter = require("devtools/shared/event-emitter");

/**
 * Shortcuts for lazily accessing and setting various preferences.
 * Usage:
 *   let prefs = new Prefs("root.path.to.branch", {
 *     myIntPref: ["Int", "leaf.path.to.my-int-pref"],
 *     myCharPref: ["Char", "leaf.path.to.my-char-pref"],
 *     myJsonPref: ["Json", "leaf.path.to.my-json-pref"],
 *     myFloatPref: ["Float", "leaf.path.to.my-float-pref"]
 *     ...
 *   });
 *
 * Get/set:
 *   prefs.myCharPref = "foo";
 *   let aux = prefs.myCharPref;
 *
 * Observe:
 *   prefs.registerObserver();
 *   prefs.on("pref-changed", (prefValue) => {
 *     ...
 *   });
 *
 * @param string prefsRoot
 *        The root path to the required preferences branch.
 * @param object prefsBlueprint
 *        An object containing { accessorName: [prefType, prefName] } keys.
 */
function PrefsHelper(prefsRoot = "", prefsBlueprint = {}) {
  EventEmitter.decorate(this);

  const cache = new Map();

  for (const accessorName in prefsBlueprint) {
    const [prefType, prefName, fallbackValue] = prefsBlueprint[accessorName];
    map(
      this,
      cache,
      accessorName,
      prefType,
      prefsRoot,
      prefName,
      fallbackValue
    );
  }

  const observer = makeObserver(this, cache, prefsRoot, prefsBlueprint);
  this.registerObserver = () => observer.register();
  this.unregisterObserver = () => observer.unregister();
}

/**
 * Helper method for getting a pref value.
 *
 * @param Map cache
 * @param string prefType
 * @param string prefsRoot
 * @param string prefName
 * @param string|int|boolean fallbackValue
 * @return any
 */
function get(cache, prefType, prefsRoot, prefName, fallbackValue) {
  const cachedPref = cache.get(prefName);
  if (cachedPref !== undefined) {
    return cachedPref;
  }
  const value = Services.prefs["get" + prefType + "Pref"](
    [prefsRoot, prefName].join("."),
    fallbackValue
  );
  cache.set(prefName, value);
  return value;
}

/**
 * Helper method for setting a pref value.
 *
 * @param Map cache
 * @param string prefType
 * @param string prefsRoot
 * @param string prefName
 * @param any value
 */
function set(cache, prefType, prefsRoot, prefName, value) {
  Services.prefs["set" + prefType + "Pref"](
    [prefsRoot, prefName].join("."),
    value
  );
  cache.set(prefName, value);
}

/**
 * Maps a property name to a pref, defining lazy getters and setters.
 * Supported types are "Bool", "Char", "Int", "Float" (sugar around "Char"
 * type and casting), and "Json" (which is basically just sugar for "Char"
 * using the standard JSON serializer).
 *
 * @param PrefsHelper self
 * @param Map cache
 * @param string accessorName
 * @param string prefType
 * @param string prefsRoot
 * @param string prefName
 * @param string|int|boolean fallbackValue
 * @param array serializer [optional]
 */
function map(
  self,
  cache,
  accessorName,
  prefType,
  prefsRoot,
  prefName,
  fallbackValue,
  serializer = { in: e => e, out: e => e }
) {
  if (prefName in self) {
    throw new Error(
      `Can't use ${prefName} because it overrides a property` +
        "on the instance."
    );
  }
  if (prefType == "Json") {
    map(
      self,
      cache,
      accessorName,
      "String",
      prefsRoot,
      prefName,
      fallbackValue,
      {
        in: JSON.parse,
        out: JSON.stringify,
      }
    );
    return;
  }
  if (prefType == "Float") {
    map(self, cache, accessorName, "Char", prefsRoot, prefName, fallbackValue, {
      in: Number.parseFloat,
      out: n => n + "",
    });
    return;
  }

  Object.defineProperty(self, accessorName, {
    get: () =>
      serializer.in(get(cache, prefType, prefsRoot, prefName, fallbackValue)),
    set: e => {
      set(cache, prefType, prefsRoot, prefName, serializer.out(e));
    },
  });
}

/**
 * Finds the accessor for the provided pref, based on the blueprint object
 * used in the constructor.
 *
 * @param PrefsHelper self
 * @param object prefsBlueprint
 * @return string
 */
function accessorNameForPref(somePrefName, prefsBlueprint) {
  for (const accessorName in prefsBlueprint) {
    const [, prefName] = prefsBlueprint[accessorName];
    if (somePrefName == prefName) {
      return accessorName;
    }
  }
  return "";
}

/**
 * Creates a pref observer for `self`.
 *
 * @param PrefsHelper self
 * @param Map cache
 * @param string prefsRoot
 * @param object prefsBlueprint
 * @return object
 */
function makeObserver(self, cache, prefsRoot, prefsBlueprint) {
  return {
    register: function() {
      this._branch = Services.prefs.getBranch(prefsRoot + ".");
      this._branch.addObserver("", this);
    },
    unregister: function() {
      this._branch.removeObserver("", this);
    },
    observe: function(subject, topic, prefName) {
      // If this particular pref isn't handled by the blueprint object,
      // even though it's in the specified branch, ignore it.
      const accessorName = accessorNameForPref(prefName, prefsBlueprint);
      if (!(accessorName in self)) {
        return;
      }
      cache.delete(prefName);
      self.emit("pref-changed", accessorName, self[accessorName]);
    },
  };
}

exports.PrefsHelper = PrefsHelper;

/**
 * A PreferenceObserver observes a pref branch for pref changes.
 * It emits an event for each preference change.
 */
function PrefObserver(branchName) {
  this.branchName = branchName;
  this.branch = Services.prefs.getBranch(branchName);
  this.branch.addObserver("", this);

  EventEmitter.decorate(this);
}

exports.PrefObserver = PrefObserver;

PrefObserver.prototype = {
  observe: function(subject, topic, data) {
    if (topic == "nsPref:changed") {
      this.emit(this.branchName + data);
    }
  },

  destroy: function() {
    if (this.branch) {
      this.branch.removeObserver("", this);
    }
  },
};
