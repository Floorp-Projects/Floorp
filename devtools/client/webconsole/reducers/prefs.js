/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
"use strict";

const {
  EAGER_EVALUATION_TOGGLE,
  WARNING_GROUPS_TOGGLE,
  AUTOCOMPLETE_TOGGLE,
} = require("devtools/client/webconsole/constants");

const PrefState = overrides =>
  Object.freeze(
    Object.assign(
      {
        logLimit: 1000,
        sidebarToggle: false,
        groupWarnings: false,
        autocomplete: false,
        eagerEvaluation: false,
        historyCount: 50,
      },
      overrides
    )
  );

const dict = {
  [EAGER_EVALUATION_TOGGLE]: "eagerEvaluation",
  [WARNING_GROUPS_TOGGLE]: "groupWarnings",
  [AUTOCOMPLETE_TOGGLE]: "autocomplete",
};

function prefs(state = PrefState(), action) {
  const pref = dict[action.type];
  if (pref) {
    return {
      ...state,
      [pref]: !state[pref],
    };
  }

  return state;
}

module.exports = {
  PrefState,
  prefs,
};
