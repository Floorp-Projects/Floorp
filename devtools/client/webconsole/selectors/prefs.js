/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
"use strict";

function getAllPrefs(state) {
  return state.prefs;
}

function getLogLimit(state) {
  return state.prefs.logLimit;
}

module.exports = {
  getAllPrefs,
  getLogLimit,
};
