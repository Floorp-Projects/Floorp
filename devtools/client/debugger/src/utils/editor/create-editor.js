/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import SourceEditor from "./source-editor";
import { features, prefs } from "../prefs";

export function createEditor() {
  const gutters = ["breakpoints", "hit-markers", "CodeMirror-linenumbers"];

  if (features.codeFolding) {
    gutters.push("CodeMirror-foldgutter");
  }

  return new SourceEditor({
    mode: "javascript",
    foldGutter: features.codeFolding,
    enableCodeFolding: features.codeFolding,
    readOnly: true,
    lineNumbers: true,
    theme: "mozilla",
    styleActiveLine: false,
    lineWrapping: prefs.editorWrapping,
    matchBrackets: true,
    showAnnotationRuler: true,
    gutters,
    value: " ",
    extraKeys: {
      // Override code mirror keymap to avoid conflicts with split console.
      Esc: false,
      "Cmd-F": false,
      "Ctrl-F": false,
      "Cmd-G": false,
      "Ctrl-G": false,
    },
    cursorBlinkRate: prefs.cursorBlinkRate,
  });
}

export function createHeadlessEditor() {
  const editor = createEditor();
  editor.appendToLocalElement(document.createElement("div"));
  return editor;
}
