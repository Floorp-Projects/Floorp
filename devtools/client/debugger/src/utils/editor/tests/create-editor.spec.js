/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { createEditor } from "../create-editor";
import SourceEditor from "../source-editor";

import { features } from "../../prefs";

describe("createEditor", () => {
  test("Returns a SourceEditor", () => {
    const editor = createEditor();
    expect(editor).toBeInstanceOf(SourceEditor);
    expect(editor.opts).toMatchSnapshot();
    expect(editor.opts.gutters).not.toContain("CodeMirror-foldgutter");
  });

  test("Adds codeFolding", () => {
    features.codeFolding = true;
    const editor = createEditor();
    expect(editor).toBeInstanceOf(SourceEditor);
    expect(editor.opts).toMatchSnapshot();
    expect(editor.opts.gutters).toContain("CodeMirror-foldgutter");
  });
});
