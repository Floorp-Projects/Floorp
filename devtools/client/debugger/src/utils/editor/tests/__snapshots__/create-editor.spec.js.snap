// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createEditor Adds codeFolding 1`] = `
Object {
  "cursorBlinkRate": 530,
  "enableCodeFolding": true,
  "extraKeys": Object {
    "Cmd-F": false,
    "Cmd-G": false,
    "Ctrl-F": false,
    "Ctrl-G": false,
    "Esc": false,
  },
  "foldGutter": true,
  "gutters": Array [
    "breakpoints",
    "hit-markers",
    "CodeMirror-linenumbers",
    "CodeMirror-foldgutter",
  ],
  "lineNumbers": true,
  "lineWrapping": false,
  "matchBrackets": true,
  "mode": "javascript",
  "readOnly": true,
  "showAnnotationRuler": true,
  "styleActiveLine": false,
  "theme": "mozilla",
  "value": " ",
}
`;

exports[`createEditor Returns a SourceEditor 1`] = `
Object {
  "cursorBlinkRate": 530,
  "enableCodeFolding": false,
  "extraKeys": Object {
    "Cmd-F": false,
    "Cmd-G": false,
    "Ctrl-F": false,
    "Ctrl-G": false,
    "Esc": false,
  },
  "foldGutter": false,
  "gutters": Array [
    "breakpoints",
    "hit-markers",
    "CodeMirror-linenumbers",
  ],
  "lineNumbers": true,
  "lineWrapping": false,
  "matchBrackets": true,
  "mode": "javascript",
  "readOnly": true,
  "showAnnotationRuler": true,
  "styleActiveLine": false,
  "theme": "mozilla",
  "value": " ",
}
`;
